### PR TITLE
Enhancement: Implement LazyListener

### LINKED ISSUE #59: Question: Can the Emitter be made ContainerAware?
How about having the `Emitter` implement the `ContainerAware` interface, so we can lazily fetch listeners from the container when we need them, instead of having to wire up a bunch of listeners which may never be needed for the current request?

```php
class ContainerAwareEmitter implements ContainerAware
{
    protected function ensureListener($listener)
    {
        if ($listener instanceof ListenerInterface) {
            return $listener;
        }

        if (is_callable($listener)) {
            return CallbackListener::fromCallable($listener);
        }
        
        if (is_string($listener)) {
            return LazyListener::fromAlias($this->container, $listener);
        }

        throw new InvalidArgumentException('Listeners should be be ListenerInterface, Closure or callable, or an alias. Received type: '.gettype($listener));
    }
}

class LazyListener implements ListenerInterface
{
    /**
     * @var ContainerInterface
     */
    protected $container;

    /**
     * @var string
     */
    protected $alias;

    /**
     * @var ListenerInterface
     */
    protected $listener;

    /**
     * @param ContainerInterface $container
     * @param string             $alias
     */
    public function __construct(ContainerInterface $container, $alias)
    {
        $this->container = $container;
        $this->alias = $alias;
    }

    /**
     * Get the listener.
     *
     * @return ListenerInterface
     */
    public function getListener()
    {
        if ($this->listener === null) {
            try {
                $listener = $this->container->get($this->alias);
            } catch (\Exception $exception) {
                throw new \BadMethodCallException('Unable to fetch listener from container');
            }

            if (!$listener instanceof ListenerInterface) {
                throw new \BadMethodCallException('Fetched listener does not implement ListenerInterface');
            }

            $this->listener = $listener;
        }

        return $this->listener;
    }

    /**
     * {@inheritdoc}
     */
    public function handle(EventInterface $event)
    {
        $this->getListener()->handle($event);
    }

    /**
     * {@inheritdoc}
     */
    public function isListener($listener)
    {
        if ($listener instanceof LazyListener) {
            $listener = $listener->getListener();
        }

        return $this->listener === $listener;
    }

    /**
     * @param ContainerInterface $container
     * @param string             $alias
     * @return static
     */
    public static function fromAlias(ContainerInterface $container, $alias)
    {
        return new static($container, $alias);
    }
}
```

### PR DESCRIPTION
**Pull request by localheinz**
This PR

* [x] requires `container-interop/container-interop` as a development dependency and suggests its installation
* [x] implements a `LazyListener`, composing an alias and an implementation of `ContainerInterface`, from which it will attempt to fetch the actual listener, and then use it to handle the event

Fixes #59.

### CODE REVIEW

**Comment by localheinz** [comment]
@frankdejonge 

Updated the specs with help from @ciaranmcnulty, does this look any good?

**Comment by frankdejonge** [comment]
@localheinz hey :) I've looked over it, over-al I'm not against this idea, though I'm leaning towards having a separate package for this listener type. Also, I think the `isListener` implementation is not correct, or at least, I think it has a very unwanted side-effect. When a listener is unregistered, the listener will be check for equality, when equal removed. This means if _any_ other listener is unregistered all the lazy listeners are instantiated. If there's a need for one (or possibly more) lazy listeners and 1 listener is removed it'll have a possible O(N) hit.

**Comment by localheinz** [comment]
@frankdejonge 

Good point!

Give me a moment!

File: src/LazyListener.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    public function isListener($listener)
+    {
+        if ($listener instanceof LazyListener) {
+            return $this === $listener;
+        }
+
+        if ($this->listener !== null) {
+            return $this->listener === $listener;
```
Comment (localheinz) [inline src/LazyListener.php]: @frankdejonge 

How about this now?

**Comment by localheinz** [comment]
@frankdejonge 

Ugh, Scrutinizer!

I just tagged a small package, right here: https://github.com/refinery29/league-lazy-event/releases/tag/0.1.0

/cc @burriko

**Pull request closed without merging**